surveyor/observation.go

```diff
@@ -0,0 +1,188 @@
+		Help: "Time spent serving the request in the service",
+	}, []string{"service"})
+
+	totalLatency = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "nats_latency_total_duration",
+		Help: "Total time spent serving a service including network overheads",
+	}, []string{"service"})
+
+	requestorRTT = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "nats_latency_requestor_rtt",
+		Help: "The RTT to the client making a request",
+	}, []string{"service"})
+
+	responderRTT = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "nats_latency_responder_rtt",
+		Help: "The RTT to the service serving the request",
+	}, []string{"service"})
+
+	systemRTT = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Name: "nats_latency_system_rtt",
+		Help: "TODO",
+	}, []string{"service"})
+)
+
+func init() {
+	prometheus.MustRegister(invalidObservationsReceived)
+	prometheus.MustRegister(observationsReceived)
+	prometheus.MustRegister(serviceLatency)
+	prometheus.MustRegister(totalLatency)
+	prometheus.MustRegister(requestorRTT)
+	prometheus.MustRegister(responderRTT)
+	prometheus.MustRegister(systemRTT)
+	prometheus.MustRegister(observationsGauge)
+}
+
+// NewServiceObservation creates a new performance observation listener
+func NewServiceObservation(f string, sopts Options) (*ServiceObsListener, error) {
+	js, err := ioutil.ReadFile(f)
+	if err != nil {
+		return nil, err
+	}
+
+	opts := &serviceObsOptions{}
+	err = json.Unmarshal(js, opts)
+	if err != nil {
+		return nil, fmt.Errorf("invalid service observation configuration: %s: %s", f, err)
+	}
+
+	err = opts.Validate()
+	if err != nil {
+		return nil, fmt.Errorf("invalid service observation configuration: %s: %s", f, err)
+	}
+
+	sopts.Name = fmt.Sprintf("%s (observing %s)", sopts.Name, opts.ServiceName)
+	sopts.Credentials = opts.Credentials
+	nc, err := connect(&sopts)
+	if err != nil {
+		return nil, fmt.Errorf("nats connection failed: %s", err)
+	}
+
+	return &ServiceObsListener{
+		nc:    nc,
+		opts:  opts,
+		sopts: &sopts,
+	}, nil
+}
+
+// Start starts listening for observations
+func (o *ServiceObsListener) Start() error {
+	_, err := o.nc.Subscribe(o.opts.Topic, o.observationHandler)
+	if err != nil {
+		return fmt.Errorf("could not subscribe to observation topic for %s (%s): %s", o.opts.ServiceName, o.opts.Topic, err)
+	}
+
+	observationsGauge.Inc()
+	log.Printf("Started observing stats on %s for %s", o.opts.Topic, o.opts.ServiceName)
+
+	return nil
+}
+
+func (o *ServiceObsListener) observationHandler(m *nats.Msg) {
+	obs := &server.ServiceLatency{}
+	err := json.Unmarshal(m.Data, obs)
+	if err != nil {
+		invalidObservationsReceived.WithLabelValues(o.opts.ServiceName).Inc()
+		log.Printf("Unparsable observation received on %s: %s", o.opts.Topic, err)
+		return
+	}
+
+	observationsReceived.WithLabelValues(o.opts.ServiceName).Inc()
+	serviceLatency.WithLabelValues(o.opts.ServiceName).Observe(obs.ServiceLatency.Seconds())
```

Comment:
In the statz collector, I used .UnixNano() for time, should probably be consistent.

Reply:
This is effectively equivelant of https://github.com/prometheus/client_golang/blob/333f01cef0d61f9ef05ada3d94e00e69c8d5cdda/prometheus/timer.go#L51 because typical prom use is:

```
obs := prometheus.NewTimer(thing)
defer obs.ObserverDuration()
```

but since we dont want to observe time passed rather record a duration manually we have to pass into it what it expects, it appears to expect seconds as float seems safest to do identical to what they do in theirs

Reply:
OK, makes sense.